### PR TITLE
avoid memory copy in last stage

### PR DESCRIPTION
**Pull request by AdrShv**
remove memory copy in last stage if we don't perform transpose

### CODE REVIEW

File: src/provider_interface.hpp [hunk 1]
```diff
@@ -41,7 +41,7 @@ class nervana::provider_interface
 
     virtual ~provider_interface() {}
     virtual void provide(int                           idx,
-                         nervana::encoded_record_list& in_buf,
```
Comment (sfraczek) [inline src/provider_interface.hpp:44]: looks like a tiny formatting error

File: src/batch_decoder.cpp [hunk 2]
```diff
@@ -60,16 +65,16 @@ void batch_decoder::process(const int index)
     if (m_deterministic_mode)
         get_thread_local_random_engine() = m_random[index];
 
-    m_provider->provide(index, *m_inputs, *m_outputs);
+    m_provider->provide(index % m_batch_size, m_inputs->record(index), (*m_outputs)[index / m_batch_size]);
```
Comment (sfraczek) [inline src/batch_decoder.cpp:68]: reminder - please add a task about refactoring it a bit. ;)

File: src/batch_iterator.cpp [hunk 3]
```diff
@@ -112,47 +112,35 @@ fixed_buffer_map* batch_iterator_fbm::filler()
...
-        {
-            rc = nullptr;
-            break;
-        }
-
-        const string first_name         = (m_input_ptr->get_names())[0];
-        size_t       input_size         = ((*m_input_ptr)[first_name])->get_item_count();
-        size_t       current_input_size = input_size - m_src_index;
-        size_t move_count = (current_input_size <= remainder) ? current_input_size : remainder;
+    if (m_input_ptr == nullptr)
```
Comment (sfraczek) [inline src/batch_iterator.cpp]: What do you think about moving this if statement inside the if statement above?

**Review by sfraczek — APPROVED** [review]
LGTM! :)

File: src/batch_decoder.cpp (hunk 2 above)
Comment (AdrShv) [reply src/batch_decoder.cpp:68]: Yes, I did that

File: src/provider_interface.hpp (hunk 1 above)
Comment (AdrShv) [reply src/provider_interface.hpp:44]: fixed

File: src/batch_iterator.cpp (hunk 3 above)
Comment (AdrShv) [reply src/batch_iterator.cpp]: ok

**Pull request merged**